File: 백준/Silver/15649. N과 M （1）/N과 M （1）.py
import sys
input=sys.stdin.readline
n, m = map(int,input().split())
visited = [0 for _ in range(n + 1)]
arr = []


def recur(number):
    if number == m:
        print(*arr)
        return
    for i in range(1, n + 1):
        if visited[i] == 1:
            continue
        visited[i] = 1
        arr.append(i)
        recur(number + 1)
        arr.pop()
        visited[i] = 0


recur(0)

File: 백준/Silver/15649. N과 M （1）/N과 M （1）.java
import java.util.*;
class Main{
    static boolean[] visited;
    static int[] arr;
    static int n;
    static int m;
    public static void main(String[] args){
        Scanner sc = new Scanner(System.in);
        n =sc.nextInt();
        m =sc.nextInt();
        arr= new int[m];
        visited = new boolean[n];
        recur(0);
    }
    static void recur(int depth){
        if(depth ==m){
            for(int val : arr){
                System.out.print(val+" ");
            }
            System.out.println();
            return;
        }
        for(int i = 0 ; i<n;i++){
            if(!visited[i]){
                visited[i]=true;
                arr[depth]=i+1;
                recur(depth+1);
                visited[i]=false;
            }
            
        }
    }
}